packages/plugins/graphql/server/src/services/builders/type.ts

```diff
@@ -272,12 +328,39 @@ export default (context: Context) => {
 
       return objectType({
         name,
-
         definition(t) {
+          if (
+            modelType !== 'component' &&
+            isNotDisabled(contentType)('id') &&
+            strapi.plugin('graphql').config('v4CompatibilityMode', false)
+          ) {
+            t.nonNull.id('id', {
+              deprecation: 'Use `documentId` instead',
+            });
+          }
+
           if (modelType === 'component' && isNotDisabled(contentType)('id')) {
             t.nonNull.id('id');
           }
 
+          if (modelType !== 'component' && isNotDisabled(contentType)('documentId')) {
+            t.nonNull.id('documentId');
+          }
+
+          if (strapi.plugin('graphql').config('v4CompatibilityMode', false)) {
+            t.nonNull.field('attributes', {
+              deprecation: 'Use root level fields instead',
+              type: name,
+              resolve: (parent) => parent,
+            });
+
+            t.nonNull.field('data', {
```

Comment:
data isn't a restricted field though, won't that cause some issues when the v4 compat mod is enabled?

Reply:
Yes that's one of the limitations but acceptable for easing the migration